SPECv2.md

```diff
@@ -0,0 +1,99 @@
+little-endian form as this is the byte ordering used on the computer
+platforms we use.
+
+# Compression
+
+The data format allows for compression of thermal video frames
+internally and is then also compressed again using standard gzip
+compression to obtain higher levels of compression. In order to read
+data the entire file/stream must be passed through a gzip decompressor
+first.
+
+# Data Format
+
+## Field Represention
+
+Parts of the data format include variable length sections of
+fields. To allow for the addition of extra fields in the future
+without breaking existing readers, each field includes its length and
+a identifying field code. Readers should skip over fields which they
+don't recognise.
+
+A field is represented as follows:
+
+| Name   | Length | Type  | Description
+| ------ | ------ | ----- |----------------------------------------------
+| Length | 1      | uint8 | Length of the field's data
+| Code   | 1      | char  | Character identifying
+| Data   | ?      | ?     | Length & content depends on Length & Code
+
+## Identification
+
+The data format always starts with:
+
+* 4 magic bytes: "CPTV"
+* 1 byte: version code: 2
+
+## Header
+
+A single header should come next. It starts with:
+* 1 byte indicating the header: "H"
+* 1 byte indicating the number of fields in the header.
+
+After these a number of fields will exist.
+
+### Compulsory Header Fields
+
+| Name          | Length   | Code  | Type    | Description
+| ------------  | ------   | ----- | ------- | ---------------------------------------------
+| Timestamp     | 8        | 'T'   | uint64  | Microseconds since 1970-01-01 UTC
+| X resolution  | 4        | 'X'   | uint32  | Frame X resolution (columns)
+| Y resolution  | 4        | 'Y'   | uint32  | Frame Y resolution (rows)
+| Compression   | 1        | 'C'   | uint8   | Compression scheme in use (0 = uncompressed)
+| Device name   | Variable | 'D'   | string  | Device name e.g. ("somewhere01")
+
+### Optional header fields
+
+| Name          | Length   | Code  | Type    | Description
+| ------------  | ------   | ----- | ------- | ---------------------------------------------
+| Motion config | Variable | 'M'   | string  | Motion detection configuration in YAML
+| Preview secs  | 1        | 'P'   | uint8   | Number of seconds of recording before motion event was detected
+
+## Frames
+
+One or more frames will follow the header. Each frame starts with:
+* 1 byte indicating a frame: "F"
+* 1 byte indicating the number of fields in the frame.
+
+### Compulsory Frame Fields
+
+The following frame fields must exist in every frame:
+
+| Name          | Length | Code  | Type      | Description
+| ----------    | ------ | ----- | --------- | ------------------------------------------------------------------
+| Time on       | 4      | 't'   | uint32    | Time in ms since the camera was powered on
+| Bit width     | 1      | 'w'   | uint8     | Bit width of the frame data
```

Comment:
Does Bit width change for different frames?

Reply:
Yes it does. The compression algorithm uses the minimum number of bits that it can for each pixel in each frame. If less changes between adjacent frames fewer bits can be used.